server/limits.go

```diff
@@ -0,0 +1,139 @@
+package server
+
+import (
+	"context"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"go.uber.org/atomic"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/stats"
+	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/tap"
+)
+
+type GrpcMethodLimiter interface {
+	// RPCCallStarting is called before request has been read into memory.
+	// All that's known about the request at this point is grpc method name.
+	// Returned error should be convertible to gRPC Status via status.FromError,
+	// otherwise gRPC-server implementation-specific error will be returned to the client (codes.PermissionDenied in grpc@v1.55.0).
+	RPCCallStarting(methodName string) error
+	RPCCallFinished(methodName string)
+}
+
+// Custom type to hide it from other packages.
+type grpcLimitCheckContextKey int
+
+// Presence of this key in the context indicates that inflight request counter was increased for this request, and needs to be decreased when request ends.
+const (
+	requestFullMethod grpcLimitCheckContextKey = 1
+
+	errTooManyInflightRequestsMsg = "too many inflight requests"
+)
+
+var errTooManyInflightRequests = status.Error(codes.Unavailable, errTooManyInflightRequestsMsg)
+
+func newGrpcLimitCheck(maxInflight int, methodLimiter GrpcMethodLimiter, namespace string, reg prometheus.Registerer) *grpcLimitCheck {
```

Comment:
[nit] Have you considered calling it "... inflight limit" instead of just "limit"?